### PR TITLE
Expose AlternativeRouteMetadata and vanish alternatives until deviation point

### PR DESCRIPTION
**Pull request by LukasPaczos**
### Description

Opening as a draft while I work on docs and tests (there are some debug logs that you can ignore as well). This branch also depends on eventual bump of Nav Native, currently based on a snapshot.

This PR introduces `AlternativeRouteMetadata` (and its children) that describe metadata of an alternative route, this includes the precise geometry indices of the fork with the primary, as well as summed duration and distance values.

These values allow us to precisely determine the point until which the alternative route should be vanished to avoid and overlap with the primary route which had a side effect of resurfacing when the vanishing route line feature was enabled.

_Before_:

https://user-images.githubusercontent.com/16925074/161552866-6f1ba0ef-d0d8-4ad1-a925-52e4442f5e7e.mp4

_After_:

https://user-images.githubusercontent.com/16925074/161552830-9ba6dda9-6abe-40c6-88db-3dd9f9149fe4.mp4

#### How

The alternative metadata is exposed via `MapboxNavigation#getAlternativeMetadataFor(NavigationRoute)` function. The idea is that each `NavigationRoute` reference returned by Nav SDK via `RouteObserver` (excluding the primary route) and `NavigationRouteAlternativesObserver` will have alternative metadata available via the getter function at the point in time when it's delivered to the developer - this is our guarantee.

In order to be able to obtain the alternative route metadata, we first need to round-trip it through the native navigator to establish the relationship and do the calculations. This pushes the timing of the observers (especially `RouteObserver`) significantly. Previously, the `RouteObserver` was called nearly immediately and only after that routes were set to the native navigator. Now, we're reversing the flow - we first set the routes to the native navigator, capture and cache metadata results, and only then call the `RouteObserver` to guarantee that the metadata will be accessible if available.

A side effect of this is that we need to wait for all the routes to be successfully set to the native navigator before delivering `RouteProgress`, not only the primary route. Since we don't want to deliver `RouteProgress` updates before `RoutesObserver` update was dispatched, we are adding a delay before the first `RouteProgress` update is actually delivered that is equal to the time needed to process alternative routes.

#### Next iterations
There are a couple of problems in the current proposal that I'd like to iterate on and follow up separately:
1. The fact that we need to wait for alternatives to be processed before dispatching `RouteProgress` also bears a risk of skipping some initial progress updates altogether (if they are returned while alternatives are being processed). My idea is to add a channel that will buffer updates received during the alternatives processing and deliver them sequentially after the routes are processed.
2. Since the `RoutesObserver` now wait for native navigator, we can actually refactor the codebase to only return back routes which were successfully processed by Nav Native, and remove the ones that failed.
3. As you can see in the videos, there's a problem with the route geometries briefly flashing, especially during the manual route switch. This problems seems to be reproducible on `main` as well, just much less pronounced since the route does not vanish in the first place.
4. The new `AlternativeRouteMetadata` goes out-of-date whenever `RoutesObserver` or `NavigationRouteAlternativesObserver` fires and is documented as such but I'm wondering whether we should bake in a `isValid` flag into the metadata instances to better surface this state. Additionally, this open up a question whether `NavigationRouteAlternativesObserver` should invalidate the metadata at all, it might just be need to invalidate metadata when `MapboxNavigation#setRoutes` is called and the primary route changes.

cc @SiarheiFedartsou @DenisPryt

### CODE REVIEW

**Comment by LukasPaczos** [comment]
There are still a couple of small things here and there I need to polish but wanted to open it up earlier for feedback on the idea itself as described.

**Comment by LukasPaczos** [comment]
> The fact that we need to wait for alternatives to be processed before dispatching RouteProgress also bears a risk of skipping some initial progress updates altogether (if they are returned while alternatives are being processed). My idea is to add a channel that will buffer updates received during the alternatives processing and deliver them sequentially after the routes are processed.

This effectively regresses https://github.com/mapbox/mapbox-navigation-android/issues/4727 but we cannot wait for the session to be started to set routes to navigator anymore (as it was introduced in https://github.com/mapbox/mapbox-navigation-android/pull/4915) because we rely on navigator's route processing even without active guidance.

**Comment by LukasPaczos** [comment]
This PR is ready for review. It's currently blocked by a need of a new Nav Native release (points to a snapshot) but code will not change upon upgrade.

The changes are not perfect just yet - there are a couple of follow-ups listed above but I wouldn't consider them blockers but additional iterations during v2.5 pre-release series, some of which could be considered blockers for a stable release though. The feature itself complete enough for us to start testing downstream.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt]: Done already!

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/AlternativeRouteInfo.kt
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/AlternativeRouteInfo.kt:20]: Argh, the AS auto-generation :face_exhaling:

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:797]: I tried explaining this in `RouteAlternativesController#paused`, could you take a look and let me know whether that's enough or not?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt:114]: Yes, that should be okay. `MapboxNavigation` does not wait for a response, only updates routes when refresh happened successfully.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
Comment (RingerJK) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:797]: yep, it's clear 👍 

IMO it's fine that users receive alternatives for the current location, we might mention it in docs. 

Additionally, sdk might skip an alternative that NN proposes and it's not in the original list.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt
Comment (RingerJK) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt:122]: Sorry, I'm blind, out of current PR scope, but important: need to add `routeIndex` and `id` to equals and hashCode functions, otherwise, they seem collisions prone

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:797]: Sorry, I don't follow the last two points, could you rephrase?

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt
Comment (RingerJK) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt:122]: han, nice suggestion! 👍

Comment (LukasPaczos) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt:122]: Done! Also update `toString` to be more readable.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
Comment (RingerJK) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:797]: > That's not ideal though because the integration pattern is to reset anything that you get in NavigationRouteAlternativesObserver back to MapboxNavigation#setRoutes, so the RoutesObserver call will effectively be duplicated. 

got it 👍 That makes sense to me

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt
Comment (LukasPaczos) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt:122]: Actually had to re-add `directionsRoute` to `equals` because we're changing annotations under the hood without updating the ID and these routes will not be equal anymore.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/AlternativeRouteIntersection.kt
Comment (VysotskiVadim) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/AlternativeRouteIntersection.kt:59]: Another crazy 🤯  idea which can help avoid implementing data classes manually: 
```suggestion

internal data class AlternativeRouteIntersectionInternal(
    val location: Point,
    val geometryIndexInRoute: Int,
    val geometryIndexInLeg: Int,
    val legIndex: Int,
) {
    fun toPublic() = AlternativeRouteIntersection(this)
}


class AlternativeRouteIntersection internal constructor(
    private val internal: AlternativeRouteIntersectionInternal
) {

    val location: Point get() = internal.location
    val geometryIndexInRoute  get() = internal.geometryIndexInRoute
    val geometryIndexInLeg  get() = internal.geometryIndexInLeg
    val legIndex  get() = internal.geometryIndexInLeg
    
    override fun equals(other: Any?) = internal.equals(other)
    override fun hashCode() = internal.hashCode()
    override fun toString() = internal.toString()
}
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/RouteAlternativesController.kt
Comment (VysotskiVadim) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/RouteAlternativesController.kt:169]: How can this happen? As I see NN tracks current route without our help. Why do they trigger alternatives update if they're already processing a new route?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/RouteAlternativesController.kt
Comment (VysotskiVadim) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/RouteAlternativesController.kt:214]: Out of scope of this PR: 
as we're serialising JSON from a string in RAM, maybe it's more about computation and we should use computation thread pool `Dispathers.Default`?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt
Comment (VysotskiVadim) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:121]: so, does it mean that NN is ready to process alternatives before it finished processing a primary route?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt
Comment (VysotskiVadim) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:114]: as `setRoutes` has become a suspend function I guess we can stop using `treadController` and start a new coroutine from the current scope:
 ```suggestion
                val updateAlternativesJob = launch(
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt
Comment (VysotskiVadim) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:121]: oh, we already do this.

Comment (VysotskiVadim) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:121]: Sorry for double checking, but @DenisPryt can you please confirm that NN is ready to update alternatives while primary route update is in progress?

**Comment by VysotskiVadim** [comment]
> The idea is that each NavigationRoute reference returned by Nav SDK via RouteObserver (excluding the primary route) and NavigationRouteAlternativesObserver will have alternative metadata available via the getter function at the point in time when it's delivered to the developer - this is our guarantee.

@LukasPaczos, what if the SDK provides route alternatives metadata later, i.e. not immediately after the first event from RouteObserver? 
For example:
1. user set routes: `mapboxNavigation.setRoutes(listOf(route1, route2)`
2. route observer triggers immediately
3. status updates are available before alternatives get processed 
4. route alternatives processing completes
5. sdk triggers route observer one more time, and this time metadata is available

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt
Comment (DenisPryt) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:121]: This is correct. Sequence of method calls will be saved even without waiting for complete.

**Comment by LukasPaczos** [comment]
@VysotskiVadim this was my initial implementation but I didn't truly like it - switching routes is a significant and expensive operation and it's much more optimal if the metadata is available right away. For example, our route line implementation uses the alternatives metadata to vanish the alternatives until the deviation point and when metadata is delivered asynchronously, it requires redraw of the lines which is unnecessarily inefficient. Since Nav Native slowly becomes the source of truth for the routes, especially when we add alternative metadata into the equation, I thought that now is as good of a moment as any to synchronize the Nav Native navigator with Nav SDK's route list (at least partially).

Additionally, in my opinion, if sometimes the metadata would be available when `RoutesObserver` fires and sometimes not, it would be hard for developers to implement a robust logic around this data.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt
Comment (VysotskiVadim) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:121]: @DenisPryt , just curious, does it mean that under the hood `setAlternativeRoutes` isn't processed until primary route is set?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/AlternativeRouteIntersection.kt
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/AlternativeRouteIntersection.kt:59]: Interesting idea :+1: Let's discuss separately.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:114]: `suspend` functions do not have a coroutine scope so we cannot start additional coroutines directly from the `suspend` function. I guess we could try and avoid starting new coroutines and instead just change a dispatcher when necessary if the work is done sequentially anyway, but I'd leave that to a separate refactor.

**Comment by VysotskiVadim** [comment]
@LukasPaczos , I see, that makes sense.

The sate of the Navigation SDK while NN processes alternatives scares me a bit. From the one side, we can't navigate as we don't have enough data to draw a route line so we wait until processing of alternatives completes. From the other side, NN starts sending us status updates as soon as it processed primary route. Now we have to come up with some workarounds to hold status updates until alternatives are processed. It feels the difference, in how Android SDK(process all then navigate) and NN(navigate when primary route ready) work, creates problems our of nothing. We have to transform one way to another, which is not easy given all the asynchronous nature of our processing 🙂 
What do you think about it?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/RouteAlternativesController.kt
Comment (VysotskiVadim) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/RouteAlternativesController.kt:169]: Do I understand correctly that:
1. Android SDK gets alternatives form NN in `onRouteAlternatives`
2. Android SDK sets alternatives back to NN using `setAlternativeRoutes`
3. NN triggers `onRouteAlternatives` with the same alternatives the SDK set in step 2
?

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/RouteAlternativesController.kt:169]: Yes, but how we obtained alternatives is irrelevant, the problem is that (before the changes in this PR):
1. Developers set routes to Nav SDK.
2. Nav SDK calls `RoutesObserver`.
3. Nav Native processes the routes and in the middle of processing calls `onRouteAlternatives` _with the same routes developer just set and received in `RoutesObserver`_.
4. Developer has a logic in `onRouteAlternatives` to update the routes and set them back to Nav SDK, creating an unnecessary, duplicate update.
6. `onRouteAlternatives` would not fire a second time with the same alternative so there's no endless loop, but there is one unnecessary callback.

**Comment by LukasPaczos** [comment]
This is a good observation and I completely agree that it's not perfect, I tried capturing it in the first post and there are definitely ways we can optimize this behavior that we'll need to explore before these changes go stable.

I would prefer collaborating with Nav Native team here to come up with a solid strategy going forward.

**Review by RingerJK — APPROVED** [review]
🚀

**Comment by LukasPaczos** [comment]
Nav Native v95.0.0 update is in. This is ready to be merged as soon as the CI finishes.

**Pull request merged**
